test/demo/utils/ofPuppeteer.ts

```diff
@@ -0,0 +1,137 @@
+import {Fin, Identity} from 'openfin/_v2/main';
+import {Browser, Page, JSHandle} from 'puppeteer';
+import {connect} from 'hadouken-js-adapter';
+
+import {uuidv4} from './uuidv4';
+
+declare const global: NodeJS.Global & {__BROWSER__: Browser};
+
+// Helper type. Works better with puppeteer than the builtin Function type
+type AnyFunction = (...args: any[]) => any;
+
+export type BaseWindowContext = Window & {fin: Fin};
+
+export class OFPuppeteerBrowser<WindowContext extends BaseWindowContext = BaseWindowContext> {
+    private _pageIdentityCache: Map<Page, Identity>;
+    private _identityPageCache: Map<string, Page>;
+    private _mountedFunctionCache: Map<Page, Map<Function, JSHandle>>;
+    private _browser: Browser;
+    private _ready: Promise<void>;
+
+    constructor() {
+        this._pageIdentityCache = new Map<Page, Identity>();
+        this._identityPageCache = new Map<string, Page>();
+        this._mountedFunctionCache = new Map<Page, Map<Function, JSHandle>>();
+        this._browser = global.__BROWSER__;
+        this._ready = this.registerCleanupListener();
+    }
+
+    public get browser() {
+        return this._browser;
+    }
+
+    public async getPage(identity: Identity): Promise<Page|undefined> {
+        await this._ready;
+        const idString = getIdString(identity);
+
+        // Return cached value when available
+        if (this._identityPageCache.has(idString)) {
+            return this._identityPageCache.get(idString);
+        }
+
+        const pages = await this._browser.pages();
+        for (const page of pages) {
+            const pageIdentity = await this.getIdentity(page);
+            if (pageIdentity && pageIdentity.uuid === identity.uuid && pageIdentity.name === identity.name) {
+                // Cache is updated by getIdentity, so no need to update here
+                return page;
+            }
+        }
+
+        // No pages found that match
+        return undefined;
+    }
+
+    public async executeOnWindow<
+        // tslint:disable-next-line: no-any Needed for tuple types.
+        T extends any[], R, C extends WindowContext = WindowContext>(executionTarget: Identity, fn: (this: C, ...args: T) => R, ...args: T):
```

Comment:
Will `unknown` work here? TypeScript should infer the type, rather than just allow anything.

Reply:
This was carried over from FDC3, and we also don't need the tslint:disable anymore, but for an actual answer the any is because this genuinely does take any. The `T` is the tuple of args that the given function will accept and is used mainly to make sure that the extra args passed to `executeOnWindow` match those expected by the given function. I don't think `unknown` is really meant for that type of thing.

Reply:
We can get rid of the generics by doing something like: 
```typescript
public async executeOnWindow<C extends WindowContext = WindowContext>(executionTarget: Identity, fn: (this: C, ...args: any[]) => any, ...args: Parameters<typeof fn>): Promise<ReturnType<typeof fn>>
```
but fundamentally we accept any function, so there will need to be an any somewhere